_posts/2024-02-23-enhanced-multi-vector-support-in-opensearch-knn.md

```diff
@@ -0,0 +1,191 @@
+
+For example, consider three documents labeled 1, 2, and 3, each containing two vectors in the nested k-NN field. Document 1 contains vectors A and B, Document 2 contains vectors C and F, and Document 3 contains vectors D and E. Let's say the two nearest vectors are A and B, both belonging to Document 1. When searching for the nested k-NN field, the search returns only Document 1, even if the k value is set to 2, as shown in the following image.
+
+<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-before-1.png" alt="multi-vector-behavior-before" class="img-centered"/>
+
+### Low recall
+Returning fewer results mentioned earlier can also lead to low recall. In OpenSearch, an index consists of multiple shards, each containing multiple segments. Search operates at the segment level, with results aggregated at the shard level. After this, the results from all shards are aggregated and returned to the user. Consequently, if a single segment contains all of the top k documents but returns fewer than k documents, the final results will contain k documents but will not represent the true top k documents. This is because the search algorithm operates at the segment level, potentially missing relevant documents from other segments that should have been included in the top k results.
+
+To illustrate this, consider the following example. Suppose there are 7 indexed documents, each containing either one or two vectors. Let's assume that, when searching the nested vector field, the order of vectors from nearest to farthest is as follows: A, B, C, D, E, F, G, H, I, and J. The search occurs at the segment level, and with a k value of 2, only document 1 is returned from segment 1. Consequently, the final aggregated result contains document 1 and document 3, whereas the expected result should include document 1 and document 2, as shown in the following image.
+
+<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-before-2.png" alt="multi-vector-behavior-before-cluster-level" class="img-centered"/>
+
+## Improvements in OpenSearch 2.12
+
+In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector closest to the query vector and discards the farther one. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.
```

Comment:
There will be only two vector to choose from at a time. We choose nearer one and drop the farther one and this can happen multiple times for the same document.

Reply:
Updated as 
```
In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector nearer to the query vector and discards the farther one out of the two. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.
```

Reply:
Updated as
`In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch keeps only the nearest vector to the query vector for each document. The distance value of the selected vector is then converted back to an OpenSearch search score, ultimately becoming the document's score.`